### ISSUE TITLE
Ability to clone entity via LUA commands

### ISSUE THREAD

**Issue by 3coms** [TheGameCreators/GameGuruRepo#493]
Hi
Would be nice to place on map just  let's say 1 health pack, and clone it via scriipt when needed,
A instance of the original one, inheriting all its properties
Ty :)

**Comment by AmenMoses** [comment]
You can already do this simply by hiding it when used and reusing the same entity.

**Comment by 3coms** [comment]
and you can have let's say 20 instance of the original one?
I just want to place 1 entity hidden on the map, and be able to clone that entity as much times as needed and place them wherever I want, via LUA, 
I've done something like that via LUA with 1 soldier, and re-using him 3 times, can remember where that thread is.

**Comment by AmenMoses** [comment]
For slaur3n's cowboy game we've got NPC's and collectibles that simply respawn somewhere else as needed, when you kill an enemy they drop 'loot' items which are re-used over and over again.  It's the way most real games do it.

**Comment by 3coms** [comment]
Thinking about trees, if you can place one on the map, and you can clone this tree just in the player fov area, as much trees as needed.  With this jnstances you can populate zones just changing their position via LUA, hide those than you don't need, for later use. 
All this being easy 4 everybody of course. ;)
well as easy as posible. LOL

**Comment by AmenMoses** [comment]
If you have to place a marker where the trees are to be placed why not simply place a tree there?

**Comment by 3coms** [comment]
marker need less resources than the trees, marker only store tree's coordinates info.

**Comment by AmenMoses** [comment]
A marker uses the exact same resources, the only extra resources would be the collision details for the ai, if you simply had a marker then placed a tree there at run time the AI wouldn't know about it so would attempt to walk straight through it, oh and the light mapper would likewise know nothing about it so you would be forced to use dynamic lighting only.

When you place a bunch of entities on the map the mesh and textures are only loaded once for the first one, after that all that is stored are the details specific to each entity, i.e. where it is and the orientation.   At run time the engine builds the scene, if you turn really quickly you can sometimes see entities popping into existence as they come into view.

**Comment by 3coms** [comment]
>  if you simply had a marker then placed a tree there at run time the AI wouldn't know about it so would attempt to walk straight through it,

So ai has to know all the trees pos around the map in every moment.  so why not reduce the tree's info?
light mapper should to scan all the scene before do nothing, be aware the new additions you does, and include then in the next render. The fact GG does not work nowaday like we need, does not ment in can't never do it.

> When you place a bunch of entities on the map the mesh and textures are only loaded once for the first one, after that all that is stored are the details specific to each entity, i.e. where it is and the orientation.

This sound wonderful, but so why we need to improve perfomance?

The user side to improve perfomance we already know.

There should be something like a big library where everybody goes for looking required info, a big sql databases, or something like that.
This database should be update each time you place/edit/delete entity on the map.
Relevant data being store there, for later use.

All db data being compared with your map when you open it.
That way GG know when _missing entity_ as example.

When you load a bunch of trees over your map because your forest scene required, when you are into one area of the map you are away from the others obviusly, ignore if GG is storing all these trees info although player are far away from there. Maybe store marker position in a simple plane file for later use, does not takes as much as resorces like a bunch of trees.

**Comment by AmenMoses** [comment]
I don't understand what you mean by 'reduce the trees info', the engine needs to know where to put it and how it is orientated and how big it is etc, it needs this information for every entity whether it is a tree, building, zone or marker etc.

The AI collision map is built once and then stored for the AI to use to navigate around the map, if you move stuff around at run time the AI will no longer be able to navigate around properly.

Any entities like collectible items don't need to be part of the AI collision map, in fact any active entity will not be included in it anyhow unless marked as immobile maybe, so you can reuse them if you want to like I described earlier which can improve performance a bit as you won't have as many scripts being run each frame.

**Comment by 3coms** [comment]
> the engine needs to know where to put it and how it is orientated and how big it is etc, it needs this information for every entity whether it is a tree, building, zone or marker etc.
Yes I know, so less trees, lets information the engine has to store.

I've try to explain with example:

Assume with has tree markers, or just one with properties panel to set type, size etc.
all tha info about the tree nature is already there and quite accesible for engine.
Assume we want to fill up some area with oak trees, we place marker where we want these trees when player nearby or in zone.

about these markers engine only store let's say id, position, orientation etc,  most of the other params are in properties panel setup, so emgine has all the needed information about these trees, maybe store in a single file txt or database.

But that information should be update at realtime and just when needed; maybe GG should to generate collision map just in time an as much as needed.
Generate whole collsion map and store it all the game, even if player is not there,  I don't know, perhaps it is not do-able in that way.

**Comment by AmenMoses** [comment]
Having GG suddenly build a forest when the player enters an area will have the exact opposite effect, i.e. it will make the performance worse not better.
You are talking about procedural generation techniques ('No Mans Sky' for example) which are hellishly difficult to get right.
If you want a forest in GG you just extract a tree press 'i' and paint away to your hearts content.

**Comment by Einlander** [comment]
>You can already do this simply by hiding it when used and reusing the same entity.

That's fine until the entity needs to be on screen multiple times or used in multiple places simultaneously.

My example would be creating an entity spawner. Let's say you want a bunch of health kits in a specific area. You should be able to make a script that says at this location spawn 5 medkits. It would not make sense to add however many medkits on the map somewhere.

Also without this you can't have wave shooters that increase the enemy count each round. Unless you plan to send in enemies one by one.

**Comment by AmenMoses** [comment]
Spawning on the fly requires dynamic memory allocations and various processor intensive things to occur, both things to be avoided during game play.
Having 5 med kits off map or simply hidden until needed is far more sensible.

For my 3D asteroids demo thingy I had several hundred asteroids hiding away ready for use and simply  moved them into position at runtime, then when 'destroyed' they simply get put back into the pool of hidden entities waiting to be used for the next 'wave' of asteroids.

To increase the difficulty each wave I simply added an extra asteroid from the pool.  Performance was excellent btw.

**Comment by Einlander** [comment]
Does Game Guru have instancing? If it does it should be able to have one object and close front that. It also doesn't solve the problem of spawning things in that were not in your level.

**Comment by 3coms** [comment]
> Does Game Guru have instancing?
I think does not.
I'm thinking about vegetation, specially animate ones, they are a big resources eater.
